File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
import collections

def solution(maps):
    candidate_dis=[] # 후보 값
    xs= len(maps[0])
    ys = len(maps)
    
    # 위, 왼, 아래, 오른
    dy=[1,0,-1,0]
    dx=[0,-1,0,1]

    def bfs(x,y,d):
        queue=collections.deque()
        queue.append((x,y,d))
        
        while queue:
            x,y,d = queue.popleft()
            
            if(y==ys-1 and x==xs-1):
                candidate_dis.append(d)
                continue
            else:
                for i in range(4):
                    nx=x+dx[i]
                    ny=y+dy[i]
                    if nx>=0 and nx<xs and ny>=0 and ny<ys:
                        if(maps[ny][nx]==1):
                            maps[ny][nx]=0
                            queue.append((nx,ny,d+1))
                    
    bfs(0,0,1)
        
    if len(candidate_dis)>0:
        candidate_dis.sort()
        return candidate_dis[0]
    else:
        return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    var answer = 0;
    const n=maps.length;
    const m=maps[0].length;
    
    // 상, 우, 하, 좌
    const dx=[0,1,0,-1];
    const dy=[1,0,-1,0];
    const visited=Array.from({length:n},()=>Array(m).fill(false));
    
    function bfs(){
        const q=[];
        q.push([0,0,0]);
        visited[0][0]=true;
        while (q.length!==0){
            [y,x,count] = q.shift();
            
            if(y===n-1 && x===m-1){
                answer = count+1
                return
            }
            
            for(let i=0;i<4;i++){
                const ny = y+dy[i]
                const nx = x+dx[i]
                if ( ny < n && ny >=0 && nx < m && nx >=0 && maps[ny][nx]!==0 && visited[ny][nx]==false){
                    visited[ny][nx]=true
                    q.push([ny,nx,count+1])
                }
            }
        }
    }
    
    bfs();
    
    if(answer==0){
        return -1
    }
    return answer;
}